File: weather_app_flutter/lib/models/weather_models.dart
enum WeatherTheme {
  standard,
  sunny,
  partlyCloudy,
  partlyCloudyNight,
  cloudy,
  cloudyNight,
  rainy,
  snowy,
  stormy,
  foggy,
  sunrise,
  sunset,
  sunsetGlow,
  night,
}

class CitySuggestion {
  const CitySuggestion({
    required this.id,
    required this.name,
    required this.country,
    required this.latitude,
    required this.longitude,
    required this.label,
    this.admin1,
  });

  final String id;
  final String name;
  final String country;
  final String? admin1;
  final double latitude;
  final double longitude;
  final String label;
}

class WeatherViewModel {
  const WeatherViewModel({
    required this.city,
    required this.country,
    required this.temperature,
    required this.windSpeed,
    required this.humidity,
    required this.cloudCover,
    required this.description,
    required this.iconUrl,
    required this.theme,
    required this.weatherCode,
    required this.utcOffsetSeconds,
    required this.isDaylight,
    required this.sunRise,
    required this.sunSet,
    required this.moonRise,
    required this.moonSet,
    required this.sunRiseLabel,
    required this.sunSetLabel,
    required this.moonRiseLabel,
    required this.moonSetLabel,
    required this.localDateTime,
    required this.updatedAtLabel,
  });

  final String city;
  final String country;
  final int temperature;
  final int windSpeed;
  final int humidity;
  final int cloudCover;
  final String description;
  final String iconUrl;
  final WeatherTheme theme;
  final int weatherCode;
  final int utcOffsetSeconds;
  final bool isDaylight;
  final String? sunRise;
  final String? sunSet;
  final String? moonRise;
  final String? moonSet;
  final String sunRiseLabel;
  final String sunSetLabel;
  final String moonRiseLabel;
  final String moonSetLabel;
  final String localDateTime;
  final String updatedAtLabel;

  WeatherViewModel copyWith({
    String? iconUrl,
    WeatherTheme? theme,
    String? localDateTime,
    bool? isDaylight,
  }) {
    return WeatherViewModel(
      city: city,
      country: country,
      temperature: temperature,
      windSpeed: windSpeed,
      humidity: humidity,
      cloudCover: cloudCover,
      description: description,
      iconUrl: iconUrl ?? this.iconUrl,
      theme: theme ?? this.theme,
      weatherCode: weatherCode,
      utcOffsetSeconds: utcOffsetSeconds,
      isDaylight: isDaylight ?? this.isDaylight,
      sunRise: sunRise,
      sunSet: sunSet,
      moonRise: moonRise,
      moonSet: moonSet,
      sunRiseLabel: sunRiseLabel,
      sunSetLabel: sunSetLabel,
      moonRiseLabel: moonRiseLabel,
      moonSetLabel: moonSetLabel,
      localDateTime: localDateTime ?? this.localDateTime,
      updatedAtLabel: updatedAtLabel,
    );
  }
}

File: src/app/services/weather.models.ts
export type WeatherTheme =
  | 'default'
  | 'sunny'
  | 'partly-cloudy'
  | 'partly-cloudy-night'
  | 'cloudy'
  | 'cloudy-night'
  | 'rainy'
  | 'snowy'
  | 'stormy'
  | 'foggy'
  | 'sunrise'
  | 'sunset'
  | 'sunset-glow'
  | 'night';

export type CitySuggestion = {
  id: string;
  name: string;
  country: string;
  admin1?: string;
  admin2?: string;
  countryCode?: string;
  latitude: number;
  longitude: number;
  label: string;
};

export type HourlyForecast = {
  time: string;
  timeLabel: string;
  dateLabel: string;
  temperature: number;
  apparentTemperature: number;
  humidity: number;
  precipitationProbability: number;
  precipitation: number;
  windSpeed: number;
  windGusts: number;
  weatherCode: number;
  description: string;
  iconUrl: string;
};

export type DailyPeriodForecast = {
  label: string;
  temperature: number;
  apparentTemperature: number;
  humidity: number;
  precipitationProbability: number;
  windSpeed: number;
  description: string;
  iconUrl: string;
};

export type DailyForecast = {
  date: string;
  dateLabel: string;
  shortDateLabel: string;
  weatherCode: number;
  description: string;
  iconUrl: string;
  temperatureMax: number;
  temperatureMin: number;
  apparentTemperatureMax: number;
  apparentTemperatureMin: number;
  precipitationProbability: number;
  precipitationSum: number;
  rainSum: number;
  showersSum: number;
  snowfallSum: number;
  windSpeedMax: number;
  windGustsMax: number;
  uvIndexMax: number;
  sunriseLabel: string;
  sunsetLabel: string;
  moonriseLabel: string;
  moonsetLabel: string;
  morning: DailyPeriodForecast | null;
  evening: DailyPeriodForecast | null;
  hours: HourlyForecast[];
};

export type WeatherViewModel = {
  city: string;
  country: string;
  admin1?: string;
  admin2?: string;
  countryCode?: string;
  locationLabel: string;
  latitude: number;
  longitude: number;
  temperature: number;
  windSpeed: number;
  humidity: number;
  cloudCover: number;
  description: string;
  iconUrl: string;
  theme: WeatherTheme;
  weatherCode: number;
  timeZone: string;
  isDaylight: boolean;
  sunRise: string | null;
  sunSet: string | null;
  moonRise: string | null;
  moonSet: string | null;
  sunRiseLabel: string;
  sunSetLabel: string;
  moonRiseLabel: string;
  moonSetLabel: string;
  localDateTime: string;
  updatedAtLabel: string;
  updatedAt: number;
  hourly: HourlyForecast[];
  daily: DailyForecast[];
};

export type OpenMeteoPlace = {
  id?: number;
  name: string;
  country?: string;
  country_code?: string;
  admin1?: string;
  admin2?: string;
  latitude: number;
  longitude: number;
};
